File: Samsung_SW_test/resign.py
# Problems from : https://www.acmicpc.net/problem/14501
# Samsung SW 역량 테스트

# get input
N = int(input().split(' ')[0])
table = list()
for _ in range(N):
    table.append([int(x) for x in input().split(' ')])

maximum = 0
def calculate(day, wage):
    global table, N, maximum
    if N < day:
        raise IndexError
    elif N == day:
        return day, wage
    else:
        for i in range(day, N):
            d, w = day, wage
            try:
                d = i + table[i][0]
                w += table[i][1]
                d, w = calculate(d, w)
            except IndexError:
                d, w = day, wage
                continue
            if maximum < w:
                maximum = w
        return d, w

if __name__ == "__main__":
    calculate(0, 0)
    print(maximum)

File: Samsung_SW_test/resign.c
#include <stdio.h>

// Problems from : https ://www.acmicpc.net/problem/14501
// Samsung SW 역량 테스트

int *T, *P;
int N;
int maximum = 0;

int calculate(int day, int wage){
	int i, d, w, next;
	if (N < day)
		return -1;
	else if (N == day)
		return wage;
	for (i = day; i < N; i++){
		d = i + T[i];
		w = wage + P[i];
		next = calculate(d, w);
		if (next==-1){
			w = wage;
			if (maximum < wage)
				maximum = wage;
		}
		else {
			w = next;
			if (maximum < next)
				maximum = next;
		}
	}

	return w;
}

int main(void){
	int i, t, p;
	scanf("%d", &N);
	T = (int)malloc(sizeof(int)*N);
	P = (int)malloc(sizeof(int)*N);
	for (i = 0; i < N; i++){
		scanf("%d %d", &t, &p);
		T[i] = t;
		P[i] = p;
	}

	calculate(0, 0);
	printf("%d", maximum);
	return 0;
}
